website_chd_product_configurator/controllers/controllers.py

```diff
@@ -0,0 +1,244 @@
+        return request.render('website_chd_product_configurator.conf_start', {
+            'conf_products': product_template_model.search([(
+                'chd_origin_product', '=', True)
+                ]),
+            })
+
+    # shows the options list
+    @http.route('/chd_init/<int:id>/', website = True)
+    def call_configurator(self, id = None, **form_data):
+        errormsg = ""
+        product_template_model = request.env['product.template']
+        chd_results_model = request.env['chd.product_configurator.result']
+        curr_product_id = product_template_model.browse([id])
+        pc_at_model = request.env[
+            'price.component.attribute.template']
+        all_accessories = []
+        all_attributes = {}
+        chd_dict = {
+            'origin_product_id': curr_product_id.id,
+            'partner_id': request.env.user.partner_id.id,
+            'state': 'config',
+            'quantity': form_data['quantity']
+            }
+        # dynamic and fixed size types
+        if form_data['product_id_chd_size_type'] == "fixed":
+            chd_dict['size_id'] = form_data['size']
+            chd_size = request.env['chd.size'].browse([
+                int(form_data['size'])
+                ])
+            chd_dict['width'] = chd_size.width
+            chd_dict['height'] = chd_size.height
+        else:
+            chd_dict['width'] = form_data['width']
+            chd_dict['height'] = form_data['height']
+        # if there aren't any errors, upload the image
+        message = ''
+        has_image = product_template_model.browse([
+            int(form_data['product_id'])
+            ]).chd_configurator_has_image
+        if has_image:
+            try:
+                import base64
+                fileitem = form_data['pic']
+                # add uploaded image to configurator
+                chd_dict['image'] = base64.b64encode(fileitem.stream.read())
+                chd_dict['image_filename'] = fileitem.filename
+                message = 'image uploaded successfully'
+            except:
+                message = 'Problems uploading your image, please contact us'
+        if message == '':
+            message = 'no image needed for this product'
+        for key in form_data:
+            if ('pricecomponent_' in key):
+                # if it is of type "string" the value is the index of the
+                # selection in pricecomponent_name, the id encoded in the key
+                # if it is of type numerical the value is the actual value of
+                # the pricecomponent field, the id is encoded in the key
+                if ('pricecomponent_string' in key) or ('pricecomponent_int' in key):
+                    pricecomponent_id = int(key.split('_')[3])
+                    pricecomponent_value = form_data[key]
+                # the configurator stores the attributes in a
+                # dictionary in the "attributes" field ,
+                all_attributes[
```

Comment:
shouldn't that line be one level more indented? Given it's a dictionary, the way it is doesn't seem to hurt - you'd overwrite the last key/value pair with this exact values, but it might bite you later. Think of what happens if the first key in form_data is 'pricecomponent_float'.